src/software/geom/algorithms/contains.cpp

```diff
@@ -61,13 +62,14 @@ bool contains(const Polygon& container, const Point& contained)
 bool contains(const Ray& container, const Point& contained)
 {
     Point point_in_ray_direction = container.getStart() + container.toUnitVector();
-
-    bool point_is_ray_start = contained == container.getStart();
+    bool point_is_ray_start      = contained == container.getStart();
     bool point_collinear_with_ray =
-        collinear(contained, container.getStart(), point_in_ray_direction);
-    bool point_is_in_ray_direction =
+        collinear(contained, container.getStart(), point_in_ray_direction,
+                  1000 * FIXED_EPSILON, 2 * ULPS_EPSILON_HUNDRED);
+    bool point_is_in_ray_direction = almostEqual(
         ((contained - container.getStart()).normalize() - container.toUnitVector())
-            .length() < FIXED_EPSILON;
+            .length(),
+        0, 1000 * FIXED_EPSILON, 2 * ULPS_EPSILON_HUNDRED);
```

Comment:
could you leave a quick comment why this specific usage of collinear needs to be given the epsilon and ulps manually and why the default won't work?

and are these the smallest values that work?

Reply:
The differences being compared varied each run and the smallest value that works depended on where I was running the tests from. If I ran from CLion then an fixed_epsilon of `1e-14` was enough, but running from terminal `1e-13` was needed some times. So to be cautious I used `1e-12`. 

For the ulps, most cases the ulps was under 200, but to eliminate all warnings using ulps alone would require something over 5000, so I actually wasn't sure if we should bother finding that value as fixed_epsilon was enough to remove warnings, but im open to thoughts?

Reply:
Thats reasonable I was just curious, could you leave a quick comment about the `1e-12` and the 200 ULPS, other than that LGTM :+1:

Reply:
added comment